### ISSUE TITLE
Lore always generates "Index" links even when meaningless, and even on index pages themselves.

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#10613]
|<img alt="tirkal's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| tirkal reported|
|-|-|
|Trac ID|trac#812|
|Type|defect|
|Created|2004-12-08 16:35:18Z|

Attachments:

* [patch.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/872/872a319ee7a24a0ea855777702e15aae09deb042/9c2db015e47d179fd5565f2f2144ace7e021399c.diff) (1009 bytes) - added by tehminkeh on 2007-06-21 19:21:45Z - 
* [patch2.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/872/872a319ee7a24a0ea855777702e15aae09deb042/8bba9b81382490b7a00da29d0b5f22d1baf61db7.diff) (2665 bytes) - added by tehminkeh on 2007-06-22 21:25:33Z - 
* [patch2.2.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/872/872a319ee7a24a0ea855777702e15aae09deb042/d3f6b07309c91d2728817a88cac06f3f839cbeeb.diff) (3204 bytes) - added by tehminkeh on 2007-06-22 22:26:42Z - 
* [patch2.3.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/872/872a319ee7a24a0ea855777702e15aae09deb042/bff1b5791a1e0b6b4f0ba191069b2fd7e8d2285f.diff) (3191 bytes) - added by tehminkeh on 2007-06-22 22:39:26Z - 
* [patch3.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/872/872a319ee7a24a0ea855777702e15aae09deb042/fc9569415d9230b8598c9f415fc62fba7db5f097.diff) (2798 bytes) - added by tehminkeh on 2007-06-23 02:47:57Z - 
* [patch3.2.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/872/872a319ee7a24a0ea855777702e15aae09deb042/aa623c2e63b8ee458cc05eed020464d83f23a9b9.diff) (3500 bytes) - added by tehminkeh on 2007-06-23 04:10:54Z - 
* [patch3.3.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/872/872a319ee7a24a0ea855777702e15aae09deb042/c349b6405bca5a24f598727eb54a19c84b280264.diff) (4049 bytes) - added by tehminkeh on 2007-06-23 17:22:57Z - 
* [patch3.4.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/872/872a319ee7a24a0ea855777702e15aae09deb042/fcdca3479384033de92e93d13c5dae9bd9116fb1.diff) (4051 bytes) - added by tehminkeh on 2007-06-23 17:25:28Z - 
* [patch3.5.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/872/872a319ee7a24a0ea855777702e15aae09deb042/92cdec232dce6cebce7232432d20b3c071739637.diff) (4050 bytes) - added by tehminkeh on 2007-06-23 17:28:33Z - 
* [patch3.6.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/872/872a319ee7a24a0ea855777702e15aae09deb042/7fdea5e824072b9aafa9a924d51db4e6748ef2b6.diff) (3776 bytes) - added by tehminkeh on 2007-06-23 18:13:59Z - 
* [patch3.7.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/872/872a319ee7a24a0ea855777702e15aae09deb042/df4cd26cc1e84b789aa2b1faefdc59c5e30ccad2.diff) (3940 bytes) - added by tehminkeh on 2007-06-23 18:23:09Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__812 812
type__defect defect
reporter__tirkal tirkal
priority__high high
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__lore lore
keywords__ 
time__1102523718000000 1102523718000000
changetime__1182636975000000 1182636975000000
version__ 
owner__ 
cc__radix cc__spiv cc__tirkal
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="tirkal's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|tirkal commented|
|-|-|

```
#!html
<pre>
Pages in the Twisted docs have an "Index" link at the bottom. This link is
sometimes invalid. For example, the link at the bottom of the this page:

http://twisted.sourceforge.net/TwistedDocs-1.3.0/howto/tutorial/intro.html

Leads to a non-existing file (404).

The exact same issue affects the downloadable HTML docs.

The docs on tm.com seem to be less affected. Following the index link at the
bottom of the same page there:

http://twistedmatrix.com/documents/current/howto/tutorial/intro

at least doesn't cause a 404. But it points to the HOWTOs index page:

http://twistedmatrix.com/documents/current/howto/index

instead of the more logical tutorial index page:

http://twistedmatrix.com/documents/current/howto/tutorial/index

(after all, the index link is located on a tutorial page, and since it leads
somewhere else the important direct link from the tutorial page to its index is
missing).

Similar issues were encountered throughout the docs: for example, the "Index"
link at the bottom of the HOWTOs index page points to the HOWTOs index page (ie
the link points to the same page in which it resides), while the logical link to
the next index up (the top-level documentaion index page) is missing.

Also, the "Examples" link from the top-level index points to a 404.

More generally, the docs navigation system is could use a major overhaul.
Important navigation links are missing. For example, where are the "next page"
"previous page" links?
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="puzzlement's avatar" src="https://avatars.githubusercontent.com/u/94942?s=50" width="50" height="50">](https://github.com/puzzlement)|@puzzlement commented|
|-|-|

```
#!html
<pre>
This is really two separate bugs. The first is that the doc system will always
always always generate these links: it's not something I have control over as a
documentation author: it's embedded in the system. I'll retitle this bug and
reassign it as a Lore bug.

The second is a "make a separate index page for the tutorial". I'll file it and
make you nosy.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)|@radix commented|
|-|-|

```
#!html
<pre>
I've added #1257 about the 404 broken links in the tutorial page.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun set owner to tehminkeh|
|-|-|

Patch looks pretty good, just have two comments, in order of importance:

  * there's no unit test included - ideally, code should be developed "test first": write the unit test, verify that it fails, then adjust the implementation to make it pass.  Then you know when you've actually succeeded at resolving an issue, and you never forget to test any of your code either :)  So, add a unit test for this.
  * The whitespace in the docstring is a little weird: ```index-    link``` - I guess that was just a typo or something.

**Comment by twisted-trac** [comment]
|<img alt="tehminkeh's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_9"></a>|tehminkeh commented|
|-|-|

Ok, I finally have a patch that I didn't find glaring errors in. (patch2.3)

**Comment by twisted-trac** [comment]
|<img alt="tehminkeh's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|tehminkeh removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="z3p's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_11"></a>|z3p set owner to tehminkeh|
|-|-|

This patch looks good to me.

**Comment by twisted-trac** [comment]
|<img alt="tehminkeh's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_12"></a>|tehminkeh removed owner|
|-|-|

patch3.diff and patch2.diff work exactly the same, but patch3 code is cleaner

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_13"></a>|@glyph commented|
|-|-|

tehminkeh: Before it gets merged, the test's docstring really ought to explain what behavior it is verifying, not just where the code lives.  Right now reading the test I'm not exactly sure how to read it.

z3p: To my knowledge, tehminkeh doesn't have commit access yet, so we shouldn't delete the review keyword unless there's more work to do or the ticket has been merged.  (Right now I'm deleting it because I think there is more work to do.)

Next time it pops up into review, if it's OK, someone should merge it.

**Comment by twisted-trac** [comment]
|<img alt="tehminkeh's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_14"></a>|tehminkeh commented|
|-|-|

ok, formatting has been fixed. (I hope)

**Comment by twisted-trac** [comment]
|<img alt="z3p's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_15"></a>|z3p commented|
|-|-|

This patch looks good to me.  I'm heading to bed as we speak, but if this is still uncommitted in the morning, I'll check it in.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_16"></a>|@exarkun commented|
|-|-|

Some lines in the patch are >80 cols and some lines use tabs as indentation.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_17"></a>|@exarkun set status to `closed`|
|-|-|

(In [20597]) Apply Lore index-link node handling patch

Author: Stephen Drodge and Mounika Garlapati
Reviewer: z3p, exarkun
Fixes [#10613](https://github.com/twisted/twisted/issues/10613)

Change twisted.lore.tree.setIndexLink so that it removes node with the
index-link class when the specified index filename is None, rather than
leaving them in place unmodified.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
